#  Project :- Wrangle and Analyze Data 

By Hem Bharti

## Gathering Data

In [1]:
import pandas as pd
import numpy as np
import requests
import tweepy
import os
import json
import time
import re
import matplotlib.pyplot as plt
import warnings

In [2]:
# read csv as a Pandas DataFrame
arch = pd.read_csv('/home/jovyan/demo/twitter-archive-enhanced.csv')
arch.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [3]:
arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [4]:
# Use requests library to download tsv file
url="https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(url)
with open('/home/jovyan/demo/image_predictions.tsv', 'wb') as file:
    file.write(response.content)
image_pred = pd.read_csv('/home/jovyan/demo/image_predictions.tsv', sep='\t')


In [5]:
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


**Query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file.**


In [7]:
CONSUMER_KEY = ""
CONSUMER_SECRET = ""
OAUTH_TOKEN = ""
OAUTH_TOKEN_SECRET = "" 

In [8]:
auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(OAUTH_TOKEN, OAUTH_TOKEN_SECRET)
api = tweepy.API(auth)

In [9]:
# List of the error tweets
error_list = []
# List of tweets
df_list = []
# Calculate the time of execution
start = time.time()

# For loop which will add each available tweet json to df_list
for tweet_id in arch['tweet_id']:
    try:
        tweet = api.get_status(tweet_id, tweet_mode='extended',
                               wait_on_rate_limit = True, wait_on_rate_limit_notify = True)._json 
        favorites = tweet['favorite_count'] # How many favorites the tweet had
        retweets = tweet['retweet_count'] # Count of the retweet
        user_followers = tweet['user']['followers_count'] # How many followers the user had
        user_favourites = tweet['user']['favourites_count'] # How many favorites the user had
        date_time = tweet['created_at'] # The date and time of the creation
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        error_list.append(tweet_id)
# Calculate the time of excution
end = time.time()
print(end - start)

888202515573088257 _ [{'message': 'No status found with that ID.', 'code': 144}]
873697596434513921 _ [{'message': 'No status found with that ID.', 'code': 144}]
869988702071779329 _ [{'message': 'No status found with that ID.', 'code': 144}]
866816280283807744 _ [{'message': 'No status found with that ID.', 'code': 144}]
861769973181624320 _ [{'message': 'No status found with that ID.', 'code': 144}]
845459076796616705 _ [{'message': 'No status found with that ID.', 'code': 144}]
842892208864923648 _ [{'message': 'No status found with that ID.', 'code': 144}]
837012587749474308 _ [{'message': 'No status found with that ID.', 'code': 144}]
827228250799742977 _ [{'message': 'No status found with that ID.', 'code': 144}]
802247111496568832 _ [{'message': 'No status found with that ID.', 'code': 144}]
775096608509886464 _ [{'message': 'No status found with that ID.', 'code': 144}]


Rate limit reached. Sleeping for: 330


714141408463036416 _ Failed to send request: EOF occurred in violation of protocol (_ssl.c:719)
3023.67219209671


In [12]:
# lengh of the result
print("The lengh of the result", len(df_list))
# The tweet_id of the errors
print("The lengh of the errors", len(error_list))

The lengh of the result 2344
The lengh of the errors 12


In [13]:
print("The length of the result", len(df_list))

The length of the result 2344


In [14]:
# Create DataFrames from list of dictionaries
json_tweets = pd.DataFrame(df_list, columns = ['tweet_id', 'favorites', 'retweets',
                                               'user_followers', 'user_favourites', 'date_time'])
# Save the dataFrame in file
json_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [15]:
# Read the saved tweet_json.txt file into a dataframe
tweet = pd.read_csv('tweet_json.txt', encoding = 'utf-8')
tweet

tweet_id  favorites  retweets  user_followers  \
0     892420643555336193      38952      8636         6629272   
1     892177421306343426      33352      6343         6629272   
2     891815181378084864      25127      4209         6629272   
3     891689557279858688      42300      8755         6629272   
4     891327558926688256      40458      9517         6629272   
5     891087950875897856      20292      3154         6629272   
6     890971913173991426      11899      2100         6629272   
7     890729181411237888      65802     19146         6629272   
8     890609185150312448      27870      4319         6629272   
9     890240255349198849      32054      7514         6629272   
10    890006608113172480      30764      7436         6629272   
11    889880896479866881      27891      5027         6629272   
12    889665388333682689      48276     10188         6629272   
13    889638837579907072      27279      4598         6629272   
14    889531135344209921      15140      2262         6629272   
15    889278841981685760      25392      5501         6629272   
16    888917238123831296      29190      4564         6629272   
17    888804989199671297      25692      4396         6629273   
18    888554962724278272      19997      3629         6629273   
19    888078434458587136      21848      3543         6629273   
20    887705289381826560      30303      5464         6629273   
21    887517139158093824      46409     11807         6629273   
22    887473957103951883      69376     18454         6629273   
23    887343217045368832      33815     10533         6629273   
24    887101392804085760      30668      6038         6629273   
25    886983233522544640      35288      7862         6629273   
26    886736880519319552      12124      3340         6629273   
27    886680336477933568      22519      4528         6629273   
28    886366144734445568      21198      3225         6629273   
29    886267009285017600        116         4         6629274   
...                  ...        ...       ...             ...   
2314  666411507551481857        450       333         6629797   
2315  666407126856765440        111        41         6629799   
2316  666396247373291520        168        87         6629800   
2317  666373753744588802        191        94         6629800   
2318  666362758909284353        785       579         6629800   
2319  666353288456101888        223        74         6629801   
2320  666345417576210432        301       142         6629801   
2321  666337882303524864        199        93         6629801   
2322  666293911632134144        513       359         6629801   
2323  666287406224695296        151        68         6629801   
2324  666273097616637952        178        77         6629801   
2325  666268910803644416        105        36         6629801   
2326  666104133288665088      14451      6690         6629801   
2327  666102155909144576         80        13         6629801   
2328  666099513787052032        158        70         6629801   
2329  666094000022159362        166        76         6629801   
2330  666082916733198337        120        45         6629801   
2331  666073100786774016        326       169         6629801   
2332  666071193221509120        148        62         6629802   
2333  666063827256086533        480       223         6629802   
2334  666058600524156928        113        58         6629802   
2335  666057090499244032        298       142         6629802   
2336  666055525042405380        440       257         6629802   
2337  666051853826850816       1231       859         6629803   
2338  666050758794694657        133        58         6629803   
2339  666049248165822465        109        40         6629803   
2340  666044226329800704        301       141         6629803   
2341  666033412701032449        125        44         6629803   
2342  666029285002620928        131        47         6629803   
2343  666020888022790149       2561       520      

In [16]:
tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2344 entries, 0 to 2343
Data columns (total 6 columns):
tweet_id           2344 non-null int64
favorites          2344 non-null int64
retweets           2344 non-null int64
user_followers     2344 non-null int64
user_favourites    2344 non-null int64
date_time          2344 non-null object
dtypes: int64(5), object(1)
memory usage: 110.0+ KB


## Assessing

In [17]:
# Print some random examples
arch.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
389   826476773533745153                    NaN                  NaN   
1124  730211855403241472                    NaN                  NaN   
1629  684567543613382656                    NaN                  NaN   
1555  688894073864884227                    NaN                  NaN   
284   838921590096166913                    NaN                  NaN   
1803  676948236477857792                    NaN                  NaN   
1684  681679526984871937                    NaN                  NaN   
1304  707411934438625280                    NaN                  NaN   
418   822462944365645825                    NaN                  NaN   
147   863432100342583297                    NaN                  NaN   

                      timestamp  \
389   2017-01-31 17:06:32 +0000   
1124  2016-05-11 01:44:07 +0000   
1629  2016-01-06 02:49:55 +0000   
1555  2016-01-18 01:22:00 +0000   
284   2017-03-07 01:17:48 +0000   
1803  2015-12-16 02:13:31 +0000   
1684  2015-12-29 03:33:58 +0000   
1304  2016-03-09 03:45:22 +0000   
418   2017-01-20 15:17:01 +0000   
147   2017-05-13 16:33:49 +0000   

                                                 source  \
389   <a href="http://twitter.com/download/iphone" r...   
1124  <a href="http://twitter.com/download/iphone" r...   
1629  <a href="http://twitter.com/download/iphone" r...   
1555  <a href="http://twitter.com/download/iphone" r...   
284   <a href="http://twitter.com/download/iphone" r...   
1803  <a href="http://twitter.com/download/iphone" r...   
1684  <a href="http://twitter.com/download/iphone" r...   
1304  <a href="http://twitter.com/download/iphone" r...   
418   <a href="http://twitter.com/download/iphone" r...   
147   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
389   This is Pilot. He has mastered the synchronize...                  NaN   
1124  This is Ziva. She doesn't know how her collar ...                  NaN   
1629  This is Bobby. He doesn't give a damn about pe...                  NaN   
1555  This is Hamrick. He's covered in corn flakes. ...                  NaN   
284   This is Arlo. He's officially the king of snow...                  NaN   
1803  This is Karl. Karl thinks he's slick. 6/10 sne...                  NaN   
1684  Can you spot Toby the guilty pupper? 7/10 woul...                  NaN   
1304  "I shall trip the big pupper with leash. Big p...                  NaN   
418   This is Gabe. He was the unequivocal embodimen...                  NaN   
147   This is Belle. She's never been more pupset. E...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
389                        NaN                        NaN   
1124                       NaN                        NaN   
1629                       NaN                        NaN   
1555                       NaN                        NaN   
284                        NaN                        NaN   
1803                       NaN                        NaN   
1684                       NaN                        NaN   
1304                       NaN                        NaN   
418                        NaN                        NaN   
147                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
389   https://twitter.com/dog_rates/status/826476773...                12   
1124  https://twitter.com/dog_rates/status/730211855...                11   
1629  https://twitter.com/dog_rates/status/684567543...                 4   
1555  https://twitter.com/dog_rates/status/688894073...                 7   
284   https://twitter.com/dog_rates/status/838921590...                13   
1803  https://twitter.com/dog_rates/status/676948236...                 6   
1684  https://twitter.com/dog_rates/status/681679526...                 7   
1304  https://twit

In [18]:
# Assessing the data programmaticaly
arch.info()
arch.describe()
arch['rating_numerator'].value_counts()
arch['rating_denominator'].value_counts()
arch['name'].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

None         745
a             55
Charlie       12
Oliver        11
Cooper        11
Lucy          11
Lola          10
Tucker        10
Penny         10
Bo             9
Winston        9
the            8
Sadie          8
an             7
Buddy          7
Bailey         7
Daisy          7
Toby           7
Jack           6
Jax            6
Dave           6
Oscar          6
Stanley        6
Koda           6
Scout          6
Rusty          6
Leo            6
Bella          6
Milo           6
George         5
            ... 
Cannon         1
Meatball       1
Cermet         1
by             1
Lizzie         1
Kial           1
Grizz          1
Ulysses        1
Amber          1
Tassy          1
Pawnd          1
Cheryl         1
Maxwell        1
Pancake        1
Vixen          1
Grizzwald      1
Gerbald        1
Rinna          1
Bert           1
Eugene         1
Bobb           1
Ebby           1
Ridley         1
Ester          1
Glacier        1
Kody           1
Koko           1
Skittle       

In [20]:
image_pred

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [21]:
image_pred.info()
image_pred['jpg_url'].value_counts()
image_pred[image_pred['jpg_url'] == 'https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg']

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   

      img_num      p1   p1_conf  p1_dog                p2  p2_conf  p2_dog  \
2070        2  basset  0.555712    True  English_springer  0.22577    True   

                               p3   p3_conf  p3_dog  
2070  German_short-haired_pointer  0.175219    True

In [22]:
# View number of entries for each source
arch.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [23]:
#For rating that don't follow pattern
arch[arch['rating_numerator'] > 20]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
290   838150277551247360           8.381455e+17         2.195506e+07   
313   835246439529840640           8.352460e+17         2.625958e+07   
340   832215909146226688                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
188   2017-04-22 19:15:32 +0000   
189   2017-04-22 19:05:32 +0000   
290   2017-03-04 22:12:52 +0000   
313   2017-02-24 21:54:03 +0000   
340   2017-02-16 13:11:49 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
902   2016-07-28 01:00:57 +0000   
979   2016-07-04 15:00:45 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1712  2015-12-25 21:06:00 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/downloa

In [24]:
#unusual names
arch[arch['name'].apply(len) < 3]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
56    881536004380872706                    NaN                  NaN   
393   825876512159186944                    NaN                  NaN   
446   819015337530290176                    NaN                  NaN   
449   819004803107983360                    NaN                  NaN   
553   804026241225523202                    NaN                  NaN   
583   800188575492947969                    NaN                  NaN   
649   792913359805018113                    NaN                  NaN   
679   789137962068021249                    NaN                  NaN   
686   788070120937619456                    NaN                  NaN   
759   778396591732486144                    NaN                  NaN   
775   776201521193218049                    NaN                  NaN   
801   772581559778025472                    NaN                  NaN   
820   770414278348247044                    NaN                  NaN   
852   765395769549590528                    NaN                  NaN   
876   761004547850530816                    NaN                  NaN   
978   749996283729883136                    NaN                  NaN   
1002  747885874273214464                    NaN                  NaN   
1004  747816857231626240                    NaN                  NaN   
1017  746872823977771008                    NaN                  NaN   
1025  746369468511756288                    NaN                  NaN   
1049  743222593470234624                    NaN                  NaN   
1193  717537687239008257                    NaN                  NaN   
1207  715733265223708672                    NaN                  NaN   
1340  704859558691414016                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1361  703079050210877440                    NaN                  NaN   
1362  703041949650034688                    NaN                  NaN   
1368  702539513671897089                    NaN                  NaN   
1382  700864154249383937                    NaN                  NaN   
1499  692187005137076224                    NaN                  NaN   
...                  ...                    ...                  ...   
2146  669923323644657664                    NaN                  NaN   
2153  669661792646373376                    NaN                  NaN   
2161  669564461267722241                    NaN                  NaN   
2191  668955713004314625                    NaN                  NaN   
2198  668815180734689280                    NaN                  NaN   
2204  668636665813057536                    NaN                  NaN   
2211  668614819948453888                    NaN                  NaN   
2218  668507509523615744                    NaN                  NaN   
2219  668496999348633600                    NaN                  NaN   
2222  668466899341221888                    NaN                  NaN   
2235  668171859951755264                    NaN                  NaN   
2249  667861340749471744                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2304  666983947667116034                    NaN                  NaN   
2311  666781792255496192                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2347  666057090499244032                    NaN         

In [25]:
#Orignal Tweets 
arch[arch['retweeted_status_id'].isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
30    886267009285017600           8.862664e+17         2.281182e+09   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

## Quality
*Completeness, Validity, Accuracy, Consistency => a.k.a content issues*

**twitter_archive dataset**
- in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id should be integers/strings instead of float.
- retweeted_status_timestamp, timestamp should be datetime instead of object (string).
- The numerator and denominator columns have invalid values.
- In several columns null objects are non-null (None to NaN).
- Name column have invalid names i.e 'None', 'a', 'an' and less than 3 characters.
- We only want original ratings (no retweets) that have images.
- We may want to change this columns type (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id and tweet_id) to string because We don't want any operations on them.
- Sources difficult to read.

**image_predictions dataset**
- Missing values from images dataset (2075 rows instead of 2356)
- Some tweet_ids have the same jpg_url
- Some tweets are have 2 different tweet_id one redirect to the other (Dataset contains retweets)

**tweet_data dataset**

- This tweet_id (666020888022790149) duplicated 8 times


## Tidiness
Untidy data => a.k.a structural issues
- No need to all the informations in images dataset, (tweet_id and jpg_url what matters)
- Dog "stage" variable in four columns: doggo, floofer, pupper, puppo
- Join 'tweet_info' and 'image_predictions' to 'twitter_archive'

## Cleaning


In [26]:
#copy dataframes
tweet_data_clean = tweet.copy()
twitter_archive_clean = arch.copy()
image_predictions_clean= image_pred.copy()

In [27]:
twitter_archive_clean = pd.merge(left=twitter_archive_clean,
                                 right=tweet_data_clean, left_on='tweet_id', right_on='tweet_id', how='inner')

In [28]:
twitter_archive_clean = twitter_archive_clean.merge(image_predictions_clean, on='tweet_id', how='inner')


In [29]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2068 entries, 0 to 2067
Data columns (total 33 columns):
tweet_id                      2068 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2068 non-null object
source                        2068 non-null object
text                          2068 non-null object
retweeted_status_id           75 non-null float64
retweeted_status_user_id      75 non-null float64
retweeted_status_timestamp    75 non-null object
expanded_urls                 2068 non-null object
rating_numerator              2068 non-null int64
rating_denominator            2068 non-null int64
name                          2068 non-null object
doggo                         2068 non-null object
floofer                       2068 non-null object
pupper                        2068 non-null object
puppo                         2068 non-null object
favorites                     2068 

In [30]:
# Select the columns to melt and to remain
MELTS_COLUMNS = ['doggo', 'floofer', 'pupper', 'puppo']
STAY_COLUMNS = [x for x in twitter_archive_clean.columns.tolist() if x not in MELTS_COLUMNS]


# Melt the the columns into values
twitter_archive_clean = pd.melt(twitter_archive_clean, id_vars = STAY_COLUMNS, value_vars = MELTS_COLUMNS, 
                         var_name = 'stages', value_name = 'dog_stage')
# Delete column 'stages'
twitter_archive_clean = twitter_archive_clean.drop('stages', 1)

In [31]:
print(twitter_archive_clean.dog_stage.value_counts())
print(len(twitter_archive_clean))

None       7938
pupper      222
doggo        80
puppo        24
floofer       8
Name: dog_stage, dtype: int64
8272


In [32]:
# Delete the retweets
twitter_archive_clean = twitter_archive_clean[pd.isnull(twitter_archive_clean.retweeted_status_id)]

# Delete duplicated tweet_id
twitter_archive_clean = twitter_archive_clean.drop_duplicates()

# Delete tweets with no pictures
twitter_archive_clean = twitter_archive_clean.dropna(subset = ['jpg_url'])

# small test
len(twitter_archive_clean)

2310

In [33]:
# Delete columns related to retweet we don't need anymore
twitter_archive_clean = twitter_archive_clean.drop('retweeted_status_id', 1)
twitter_archive_clean = twitter_archive_clean.drop('retweeted_status_user_id', 1)
twitter_archive_clean = twitter_archive_clean.drop('retweeted_status_timestamp', 1)

# Delete column date_time we imported from the API, it has the same values as timestamp column
twitter_archive_clean = twitter_archive_clean.drop('date_time', 1)

# small test
list(twitter_archive_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog',
 'dog_stage']

In [34]:
#Delete dog_stage duplicates
twitter_archive_clean = twitter_archive_clean.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

In [35]:
print(twitter_archive_clean.dog_stage.value_counts())
print(len(twitter_archive_clean))

None       1687
pupper      212
doggo        63
puppo        23
floofer       8
Name: dog_stage, dtype: int64
1993


In [36]:
# We will store the fisrt true algorithm with it's level of confidence
prediction_algorithm = []
confidence_level = []

# Get_prediction_confidence function:
# search the first true algorithm and append it to a list with it's level of confidence
# if flase prediction_algorthm will have a value of NaN
def get_prediction_confidence(dataframe):
    if dataframe['p1_dog'] == True:
        prediction_algorithm.append(dataframe['p1'])
        confidence_level.append(dataframe['p1_conf'])
    elif dataframe['p2_dog'] == True:
        prediction_algorithm.append(dataframe['p2'])
        confidence_level.append(dataframe['p2_conf'])
    elif dataframe['p3_dog'] == True:
        prediction_algorithm.append(dataframe['p3'])
        confidence_level.append(dataframe['p3_conf'])
    else:
        prediction_algorithm.append('NaN')
        confidence_level.append(0)

twitter_archive_clean.apply(get_prediction_confidence, axis=1)
twitter_archive_clean['prediction_algorithm'] = prediction_algorithm
twitter_archive_clean['confidence_level'] = confidence_level

In [37]:
list(twitter_archive_clean)


['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog',
 'dog_stage',
 'prediction_algorithm',
 'confidence_level']

In [38]:
# Delete the columns of image prediction information
twitter_archive_clean = twitter_archive_clean.drop(['img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'], 1)

In [39]:
list(twitter_archive_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'dog_stage',
 'prediction_algorithm',
 'confidence_level']

In [40]:
# let's concentrate on low values.. let's dig more
twitter_archive_clean.info()
print('in_reply_to_user_id ')
print(twitter_archive_clean['in_reply_to_user_id'].value_counts())
print('source ')
print(twitter_archive_clean['source'].value_counts())
print('user_favourites ')
print(twitter_archive_clean['user_favourites'].value_counts())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1993 entries, 0 to 7089
Data columns (total 18 columns):
tweet_id                 1993 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1993 non-null object
source                   1993 non-null object
text                     1993 non-null object
expanded_urls            1993 non-null object
rating_numerator         1993 non-null int64
rating_denominator       1993 non-null int64
name                     1993 non-null object
favorites                1993 non-null int64
retweets                 1993 non-null int64
user_followers           1993 non-null int64
user_favourites          1993 non-null int64
jpg_url                  1993 non-null object
dog_stage                1993 non-null object
prediction_algorithm     1993 non-null object
confidence_level         1993 non-null float64
dtypes: float64(3), int64(7), object(8)
memory usage: 295.8+ KB
in_reply_

In [41]:
# drop the following columns 'in_reply_to_status_id', 'in_reply_to_user_id', 'user_favourites'
twitter_archive_clean = twitter_archive_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'user_favourites'], 1)

In [42]:
# Clean the content of source column
twitter_archive_clean['source'] = twitter_archive_clean['source'].apply(lambda x: re.findall(r'>(.*)<', x)[0])

In [43]:
# Test
twitter_archive_clean

tweet_id                  timestamp              source  \
0     892420643555336193  2017-08-01 16:23:56 +0000  Twitter for iPhone   
1461  679877062409191424  2015-12-24 04:11:37 +0000  Twitter for iPhone   
1460  680055455951884288  2015-12-24 16:00:30 +0000           TweetDeck   
1459  680070545539371008  2015-12-24 17:00:27 +0000           TweetDeck   
1458  680085611152338944  2015-12-24 18:00:19 +0000           TweetDeck   
1456  680115823365742593  2015-12-24 20:00:22 +0000  Twitter for iPhone   
1455  680130881361686529  2015-12-24 21:00:12 +0000  Twitter for iPhone   
1453  680161097740095489  2015-12-24 23:00:17 +0000  Twitter for iPhone   
1450  680206703334408192  2015-12-25 02:01:30 +0000  Twitter for iPhone   
1449  680221482581123072  2015-12-25 03:00:14 +0000  Twitter for iPhone   
1462  679862121895714818  2015-12-24 03:12:15 +0000  Twitter for iPhone   
1448  680440374763077632  2015-12-25 17:30:01 +0000  Twitter for iPhone   
1446  680494726643068929  2015-12-25 21:06:00 +0000  Twitter for iPhone   
1445  680497766108381184  2015-12-25 21:18:05 +0000  Twitter for iPhone   
1444  680583894916304897  2015-12-26 03:00:19 +0000  Twitter for iPhone   
1443  680609293079592961  2015-12-26 04:41:15 +0000  Twitter for iPhone   
1442  680798457301471234  2015-12-26 17:12:55 +0000  Twitter for iPhone   
1441  680801747103793152  2015-12-26 17:25:59 +0000  Twitter for iPhone   
1439  680889648562991104  2015-12-26 23:15:17 +0000  Twitter for iPhone   
1438  680913438424612864  2015-12-27 00:49:49 +0000  Twitter for iPhone   
1437  680934982542561280  2015-12-27 02:15:25 +0000  Twitter for iPhone   
1436  680940246314430465  2015-12-27 02:36:20 +0000  Twitter for iPhone   
1435  680959110691590145  2015-12-27 03:51:18 +0000  Twitter for iPhone   
1434  680970795137544192  2015-12-27 04:37:44 +0000  Twitter for iPhone   
1433  681193455364796417  2015-12-27 19:22:30 +0000  Twitter for iPhone   
1463  679854723806179328  2015-12-24 02:42:51 +0000  Twitter for iPhone   
1493  678446151570427904  2015-12-20 05:25:42 +0000  Twitter for iPhone   
1492  678643457146150913  2015-12-20 18:29:43 +0000  Twitter for iPhone   
1491  678675843183484930  2015-12-20 20:38:24 +0000  Twitter for iPhone   
1490  678740035362037760  2015-12-21 00:53:29 +0000  Twitter for iPhone   
...                  ...                        ...                 ...   
5094  724004602748780546  2016-04-23 22:38:43 +0000  Twitter for iPhone   
5126  717790033953034240  2016-04-06 19:04:14 +0000  Twitter for iPhone   
5087  725786712245440512  2016-04-28 20:40:11 +0000  Twitter for iPhone   
5079  727314416056803329  2016-05-03 01:50:44 +0000  Twitter for iPhone   
5078  727524757080539137  2016-05-03 15:46:33 +0000  Twitter for iPhone   
5075  728015554473250816  2016-05-05 00:16:48 +0000  Twitter for iPhone   
5091  724405726123311104  2016-04-25 01:12:38 +0000  Twitter for iPhone   
6658  802239329049477120  2016-11-25 19:55:35 +0000  Twitter for iPhone   
7047  744995568523612160  2016-06-20 20:49:19 +0000  Twitter for iPhone   
6985  751132876104687617  2016-07-07 19:16:47 +0000  Twitter for iPhone   
6973  752519690950500352  2016-07-11 15:07:30 +0000  Twitter for iPhone   
6953  756275833623502848  2016-07-21 23:53:04 +0000  Twitter for iPhone   
6791  780931614150983680  2016-09-28 00:46:20 +0000  Twitter for iPhone   
6753  787717603741622272  2016-10-16 18:11:26 +0000  Twitter for iPhone   
6733  790946055508652032  2016-10-25 16:00:09 +0000  Twitter for iPhone   
6717  793195938047070209  2016-10-31 21:00:23 +0000  Twitter for iPhone   
6648  803773340896923648  2016-11-30 01:31:12 +0000  Twitter for iPhone   
6342  859607811541651456  2017-05-03 03:17:27 +0000  Twitter for iPhone   
6529  822872901745569793  2017-01-21 18:26:02 +0000  Twitter for iPhone   
6517  825026590719483904  2017-01-27 17:04:02 +0000  Twitter for iPhone   
6515  825535076884762624  2017-01-29 02:44:34 +0000  Twitter for iPhone   
6356  855851453814013952  2017-04-22 18:3

In [44]:
# View all occurences where there are more than one #/# in 'text' column
text_ratings_to_fix = twitter_archive_clean[twitter_archive_clean.text.str.contains( r"(\d+\.?\d*\/\d+\.?\d*\D+\d+\.?\d*\/\d+\.?\d*)")].text

text_ratings_to_fix

C:\Program Files\Anaconda3\envs\py35\lib\site-packages\ipykernel\__main__.py:2: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  from ipykernel import kernelapp as app


1400    This is Darrel. He just robbed a 7/11 and is i...
1622    Two gorgeous dogs here. Little waddling dog is...
1522    Meet Tassy &amp; Bee. Tassy is pretty chill, b...
1556    These two pups just met and have instantly bon...
1229    Meet Oliviér. He takes killer selfies. Has a d...
1224    This may be the greatest video I've ever been ...
1134    This is Socks. That water pup w the super legs...
1264    When bae says they can't go out but you see th...
1290    Meet Fynn &amp; Taco. Fynn is an all-powerful ...
1279    This is Eriq. His friend just reminded him of ...
1976    This is Kial. Kial is either wearing a cape, w...
1985    Two dogs in this one. Both are rare Jujitsu Py...
1892    Here we have Pancho and Peaches. Pancho is a C...
1929    This is Spark. He's nervous. Other dog hasn't ...
2018    These are Peruvian Feldspars. Their names are ...
2047    This is an Albanian 3 1/2 legged  Episcopalian...
1727    10/10 for dog. 7/10 for cat. 12/10 for human. ...
1829    Meet H

In [45]:
for entry in text_ratings_to_fix:
    mask = twitter_archive_clean.text == entry
    column_name1 = 'rating_numerator'
    column_name2 = 'rating_denominator'
    twitter_archive_clean.loc[mask, column_name1] = re.findall(r"\d+\.?\d*\/\d+\.?\d*\D+(\d+\.?\d*)\/\d+\.?\d*", entry)
    twitter_archive_clean.loc[mask, column_name2] = 10

In [46]:
twitter_archive_clean[twitter_archive_clean.text.isin(text_ratings_to_fix)]

tweet_id                  timestamp              source  \
1400  682962037429899265  2016-01-01 16:30:13 +0000  Twitter for iPhone   
1622  674646392044941312  2015-12-09 17:46:48 +0000  Twitter for iPhone   
1522  677314812125323265  2015-12-17 02:30:09 +0000  Twitter for iPhone   
1556  676191832485810177  2015-12-14 00:07:50 +0000  Twitter for iPhone   
1229  694352839993344000  2016-02-02 02:53:12 +0000  Twitter for iPhone   
1224  695064344191721472  2016-02-04 02:00:27 +0000  Twitter for iPhone   
1134  703356393781329922  2016-02-26 23:10:06 +0000  Twitter for iPhone   
1264  691483041324204033  2016-01-25 04:49:38 +0000  Twitter for iPhone   
1290  689835978131935233  2016-01-20 15:44:48 +0000  Twitter for iPhone   
1279  690400367696297985  2016-01-22 05:07:29 +0000  Twitter for iPhone   
1976  667544320556335104  2015-11-20 03:25:43 +0000  Twitter Web Client   
1985  667491009379606528  2015-11-19 23:53:52 +0000  Twitter Web Client   
1892  669037058363662336  2015-11-24 06:17:19 +0000  Twitter for iPhone   
1929  668537837512433665  2015-11-22 21:13:35 +0000  Twitter for iPhone   
2018  666835007768551424  2015-11-18 04:27:09 +0000  Twitter for iPhone   
2047  666287406224695296  2015-11-16 16:11:11 +0000  Twitter for iPhone   
1727  672248013293752320  2015-12-03 02:56:30 +0000  Twitter for iPhone   
1829  670434127938719744  2015-11-28 02:48:46 +0000  Twitter for iPhone   
1780  671154572044468225  2015-11-30 02:31:34 +0000  Twitter for iPhone   
872   740373189193256964  2016-06-08 02:41:38 +0000  Twitter for iPhone   
997   716439118184652801  2016-04-03 01:36:11 +0000  Twitter for iPhone   
1013  714258258790387713  2016-03-28 01:10:13 +0000  Twitter for iPhone   
963   722974582966214656  2016-04-21 02:25:47 +0000  Twitter for iPhone   
611   777684233540206592  2016-09-19 01:42:24 +0000  Twitter for iPhone   
821   747600769478692864  2016-06-28 01:21:27 +0000  Twitter for iPhone   
5823  673295268553605120  2015-12-06 00:17:55 +0000  Twitter for iPhone   
5754  674737130913071104  2015-12-09 23:47:22 +0000  Twitter for iPhone   

                                                   text  \
1400  This is Darrel. He just robbed a 7/11 and is i...   
1622  Two gorgeous dogs here. Little waddling dog is...   
1522  Meet Tassy &amp; Bee. Tassy is pretty chill, b...   
1556  These two pups just met and have instantly bon...   
1229  Meet Oliviér. He takes killer selfies. Has a d...   
1224  This may be the greatest video I've ever been ...   
1134  This is Socks. That water pup w the super legs...   
1264  When bae says they can't go out but you see th...   
1290  Meet Fynn &amp; Taco. Fynn is an all-powerful ...   
1279  This is Eriq. His friend just reminded him of ...   
1976  This is Kial. Kial is either wearing a cape, w...   
1985  Two dogs in this one. Both are rare Jujitsu Py...   
1892  Here we have Pancho and Peaches. Pancho is a C...   
1929  This is Spark. He's nervous. Other dog hasn't ...   
2018  These are Peruvian Feldspars. Their names are ...   
2047  This is an Albanian 3 1/2 legged  Episcopalian...   
1727  10/10 for dog. 7/10 for cat. 12/10 for human. ...   
1829  Meet Hank and Sully. Hank is very proud of the...   
1780  Meet Holly. She's trying to teach small human-...   
872   After so many requests, this is Bretagne. She ...   
997   This is Bluebert. He just saw that both #Final...   
1013  Meet Travis and Flurp. Travis is pretty chill ...   
963   Happy 4/20 from the squad! 13/10 for all https...   
611   "Yep... just as I suspected. You're not flossi...   
821   This is Bookstore and Seaweed. Bookstore is ti...   
5823  Meet Eve. She's a raging alcoholic 8/10 (would...   
5754  Meet Rufio. He is unaware of the pink legless ...   

                                          expanded_urls rating_numerator  \
1400  https://twitter.com/dog_rates/status/682962037...               10   
1622  https://twitter.com/dog_rates/status/674646392...                8   
1522  https://twitter.com/dog_rate

In [47]:

# View tweets with decimals in rating in 'text' column
twitter_archive_clean[twitter_archive_clean.text.str.contains(r"(\d+\.\d*\/\d+)")]

C:\Program Files\Anaconda3\envs\py35\lib\site-packages\ipykernel\__main__.py:3: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  app.launch_new_instance()


tweet_id                  timestamp              source  \
1446  680494726643068929  2015-12-25 21:06:00 +0000  Twitter for iPhone   
553   786709082849828864  2016-10-13 23:23:56 +0000  Twitter for iPhone   
40    883482846933004288  2017-07-08 00:28:19 +0000  Twitter for iPhone   
4745  778027034220126208  2016-09-20 00:24:34 +0000  Twitter for iPhone   

                                                   text  \
1446  Here we have uncovered an entire battalion of ...   
553   This is Logan, the Chow who lived. He solemnly...   
40    This is Bella. She hopes her smile made you sm...   
4745  This is Sophie. She's a Jubilant Bush Pupper. ...   

                                          expanded_urls rating_numerator  \
1446  https://twitter.com/dog_rates/status/680494726...               26   
553   https://twitter.com/dog_rates/status/786709082...               75   
40    https://twitter.com/dog_rates/status/883482846...                5   
4745  https://twitter.com/dog_rates/status/778027034...               27   

      rating_denominator    name  favorites  retweets  user_followers  \
1446                  10    None       1839       531         6629645   
553                   10   Logan      19939      6892         6629346   
40                    10   Bella      46129     10098         6629277   
4745                  10  Sophie       7176      1834         6629351   

                                              jpg_url dog_stage  \
1446  https://pbs.twimg.com/media/CXGaVxOWAAADjhF.jpg      None   
553   https://pbs.twimg.com/media/CurzvFTXgAA2_AP.jpg      None   
40    https://pbs.twimg.com/media/DELC9dZXUAADqUk.jpg      None   
4745  https://pbs.twimg.com/media/Cswbc2yWcAAVsCJ.jpg    pupper   

     prediction_algorithm  confidence_level  
1446               kuvasz          0.438627  
553            Pomeranian          0.467321  
40       golden_retriever          0.943082  
4745              clumber          0.946718

In [48]:
# Set correct numerators for specific tweets
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 883482846933004288) & (twitter_archive_clean['rating_numerator'] == 5), ['rating_numerator']] = 13.5
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 786709082849828864) & (twitter_archive_clean['rating_numerator'] == 75), ['rating_numerator']] = 9.75
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 778027034220126208) & (twitter_archive_clean['rating_numerator'] == 27), ['rating_numerator']] = 11.27
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 680494726643068929) & (twitter_archive_clean['rating_numerator'] == 26), ['rating_numerator']] = 11.26

#### Test

In [49]:
twitter_archive_clean[twitter_archive_clean.text.str.contains(r"(\d+\.\d*\/\d+)")]


C:\Program Files\Anaconda3\envs\py35\lib\site-packages\ipykernel\__main__.py:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  if __name__ == '__main__':


tweet_id                  timestamp              source  \
1446  680494726643068929  2015-12-25 21:06:00 +0000  Twitter for iPhone   
553   786709082849828864  2016-10-13 23:23:56 +0000  Twitter for iPhone   
40    883482846933004288  2017-07-08 00:28:19 +0000  Twitter for iPhone   
4745  778027034220126208  2016-09-20 00:24:34 +0000  Twitter for iPhone   

                                                   text  \
1446  Here we have uncovered an entire battalion of ...   
553   This is Logan, the Chow who lived. He solemnly...   
40    This is Bella. She hopes her smile made you sm...   
4745  This is Sophie. She's a Jubilant Bush Pupper. ...   

                                          expanded_urls rating_numerator  \
1446  https://twitter.com/dog_rates/status/680494726...            11.26   
553   https://twitter.com/dog_rates/status/786709082...             9.75   
40    https://twitter.com/dog_rates/status/883482846...             13.5   
4745  https://twitter.com/dog_rates/status/778027034...            11.27   

      rating_denominator    name  favorites  retweets  user_followers  \
1446                  10    None       1839       531         6629645   
553                   10   Logan      19939      6892         6629346   
40                    10   Bella      46129     10098         6629277   
4745                  10  Sophie       7176      1834         6629351   

                                              jpg_url dog_stage  \
1446  https://pbs.twimg.com/media/CXGaVxOWAAADjhF.jpg      None   
553   https://pbs.twimg.com/media/CurzvFTXgAA2_AP.jpg      None   
40    https://pbs.twimg.com/media/DELC9dZXUAADqUk.jpg      None   
4745  https://pbs.twimg.com/media/Cswbc2yWcAAVsCJ.jpg    pupper   

     prediction_algorithm  confidence_level  
1446               kuvasz          0.438627  
553            Pomeranian          0.467321  
40       golden_retriever          0.943082  
4745              clumber          0.946718

In [50]:
# Loop on all the texts and check if it has one of pronouns of male or female
# and append the result in a list

male = ['He', 'he', 'him', 'his', "he's", 'himself']
female = ['She', 'she', 'her', 'hers', 'herself', "she's"]

dog_gender = []

for text in twitter_archive_clean['text']:
    # Male
    if any(map(lambda v:v in male, text.split())):
        dog_gender.append('male')
    # Female
    elif any(map(lambda v:v in female, text.split())):
        dog_gender.append('female')
    # If group or not specified
    else:
        dog_gender.append('NaN')

# Test
len(dog_gender)

# Save the result in a new column 'dog_name'
twitter_archive_clean['dog_gender'] = dog_gender

In [51]:
print("dog_gender count \n", twitter_archive_clean.dog_gender.value_counts())

dog_gender count 
 NaN       1131
male       636
female     226
Name: dog_gender, dtype: int64


In [52]:
twitter_archive_clean.loc[twitter_archive_clean['prediction_algorithm'] == 'NaN', 'prediction_algorithm'] = None
twitter_archive_clean.loc[twitter_archive_clean['dog_gender'] == 'NaN', 'dog_gender'] = None
twitter_archive_clean.loc[twitter_archive_clean['rating_numerator'] == 'NaN', 'rating_numerator'] = 0
#twitter_archive_clean.loc[twitter_archive_clean['rating_denominator'] == 'NaN', 'rating_denominator'] = 0

In [54]:
twitter_archive_clean['tweet_id'] = twitter_archive_clean['tweet_id'].astype(str)
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean.timestamp)
twitter_archive_clean['source'] = twitter_archive_clean['source'].astype('category')
twitter_archive_clean['favorites'] = twitter_archive_clean['favorites'].astype(int)
twitter_archive_clean['retweets'] = twitter_archive_clean['retweets'].astype(int)
twitter_archive_clean['user_followers'] = twitter_archive_clean['user_followers'].astype(int)
twitter_archive_clean['dog_stage'] = twitter_archive_clean['dog_stage'].astype('category')
twitter_archive_clean['rating_numerator'] = twitter_archive_clean['rating_numerator'].astype(float)
twitter_archive_clean['rating_denominator'] = twitter_archive_clean['rating_denominator'].astype(float)
twitter_archive_clean['dog_gender'] = twitter_archive_clean['dog_gender'].astype('category')


In [55]:
twitter_archive_clean.dtypes

tweet_id                        object
timestamp               datetime64[ns]
source                        category
text                            object
expanded_urls                   object
rating_numerator               float64
rating_denominator             float64
name                            object
favorites                        int32
retweets                         int32
user_followers                   int32
jpg_url                         object
dog_stage                     category
prediction_algorithm            object
confidence_level               float64
dog_gender                    category
dtype: object

In [56]:
# Save clean DataFrame to csv file
twitter_archive_clean.drop(twitter_archive_clean.columns[twitter_archive_clean.columns.str.contains('Unnamed',case = False)],axis = 1)
twitter_archive_clean.to_csv('/home/jovyan/demo/twitter_archive_master.csv', encoding = 'utf-8', index=False)

In [57]:
twitter_archive_clean = pd.read_csv('/home/jovyan/demo/twitter_archive_master.csv')
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1993 entries, 0 to 1992
Data columns (total 16 columns):
tweet_id                1993 non-null int64
timestamp               1993 non-null object
source                  1993 non-null object
text                    1993 non-null object
expanded_urls           1993 non-null object
rating_numerator        1993 non-null float64
rating_denominator      1993 non-null float64
name                    1993 non-null object
favorites               1993 non-null int64
retweets                1993 non-null int64
user_followers          1993 non-null int64
jpg_url                 1993 non-null object
dog_stage               1993 non-null object
prediction_algorithm    1685 non-null object
confidence_level        1993 non-null float64
dog_gender              862 non-null object
dtypes: float64(3), int64(4), object(9)
memory usage: 249.2+ KB
